### PR TITLE
[MOB-2905]: Update Booking API data model Part 1

### PR DESCRIPTION
**Pull request by nurito-burrito**
## JIRA
[MOB-2905](https://karhoo.atlassian.net/browse/MOB-2905)

## DESCRIPTION
Update Booking API data model Part 1

## Types of changes

What types of changes does your code introduce to the project?
_Put an `x` in the boxes that apply_

- [ ] Bugfix 
- [ ] New feature 
- [ ] Tech Debt
- [ ] Breaking change 
- [ ] Documentation Update
- [x] Other: NSDK improvements

## CHECKLIST

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. This is simply a reminder of what we are going to look for before merging code._
Please review the guidelines for contributing to this repository.

- [x] All work relating to the ticket is complete (Code complete/Acceptance Criteria met)
- [x] Lint and unit tests pass locally (if appropriate)
- [ ] Added tests that prove the fix is effective or that feature works (if appropriate)
- [ ] Documentation (if appropriate)

## SCREENSHOTS
[Add screenshots if any for changes you have made to the App/UI]

### CODE REVIEW

File: KarhooSDK/Api/Request/TripSearch.swift [hunk 1]
```diff
@@ -14,21 +14,81 @@ public struct TripSearch: KarhooCodableModel {
     public let tripType: TripType?
     public let paginationRowCount: Int?
     public let paginationOffset: Int?
-
+    public let fleetID: String
```
Comment (EdwardW2) [inline KarhooSDK/Api/Request/TripSearch.swift:17]: Odd, unlike Android, TripStates is not passed through here.

Might need to add the missing one here of trip_states which should get the TripState object.

File: KarhooSDK/Api/Request/TripBooking.swift [hunk 2]
```diff
@@ -11,20 +11,17 @@ import Foundation
 public struct TripBooking: KarhooCodableModel, Equatable {
     
     public let quoteId: String
-    public let passengers: Passengers
```
Comment (EdwardW2) [inline KarhooSDK/Api/Request/TripBooking.swift:14]: Why was Passengers removed from here?

In the documentation, passengers is still passed through this object.
It is also missing:
Loyalty_programme
loyalty_points
train_time
train_number
cost_center_reference
partner_trip_id

File: KarhooSDK/Api/Response/CancelledByPayer.swift [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+//
+//  Created by Nurseda Balcioglu on 08/12/2020.
+//  Copyright © 2020 Flit Technologies Ltd. All rights reserved.
+//
+
+import Foundation
+
+public struct CancelledByPayer: KarhooCodableModel {
+    
+    public let id: String
```
Comment (EdwardW2) [inline KarhooSDK/Api/Response/CancelledByPayer.swift:13]: I assume this is the User object. Should we not align the names with the backend or Android?

File: KarhooSDKIntegrationTests/Service/Trip/BookTripMethodSpec.swift [hunk 4]
```diff
@@ -35,7 +35,7 @@ final class BookTripMethodSpec: XCTestCase {
                                 locale: "")
 
         let passengers = Passengers(additionalPassengers: 0, passengerDetails: [PassengerDetails(user: userInfo)])
-        let tripBooking = TripBooking(quoteId: "123", passengers: passengers)
```
Comment (EdwardW2) [inline KarhooSDKIntegrationTests/Service/Trip/BookTripMethodSpec.swift:38]: Same with my above comment, this needs to be added back. This is still required.

File: KarhooSDKTests/Mocks/Api/Request/TripBookingMock.swift [hunk 5]
```diff
@@ -15,32 +15,27 @@ final class TripBookingMock {
     private var tripBooking: TripBooking
 
     init() {
-        self.tripBooking = TripBooking(quoteId: "",
-                                       passengers: Passengers())
```
Comment (EdwardW2) [inline KarhooSDKTests/Mocks/Api/Request/TripBookingMock.swift:19]: Add Passengers back

File: KarhooSDKTests/Mocks/Api/Request/TripBookingMock.swift [hunk 6]
```diff
@@ -15,32 +15,27 @@ final class TripBookingMock {
...
+        self.tripBooking = TripBooking(quoteId: "")
     }
 
     func set(quoteId: String) -> TripBookingMock {
-        self.tripBooking = TripBooking(quoteId: quoteId,
```
Comment (EdwardW2) [inline KarhooSDKTests/Mocks/Api/Request/TripBookingMock.swift:23]: Same with above

File: KarhooSDKTests/Mocks/Api/Request/TripBookingMock.swift [hunk 7]
```diff
@@ -15,32 +15,27 @@ final class TripBookingMock {
...
-                                                passengers: tripBooking.passengers)
+        self.tripBooking = TripBooking(quoteId: quoteId)
         return self
     }
 
     func set(passengers: Passengers) -> TripBookingMock {
-        self.tripBooking = TripBooking(quoteId: tripBooking.quoteId,
```
Comment (EdwardW2) [inline KarhooSDKTests/Mocks/Api/Request/TripBookingMock.swift:29]: Same with above

File: KarhooSDKTests/Mocks/Api/Request/TripBookingMock.swift [hunk 8]
```diff
@@ -15,32 +15,27 @@ final class TripBookingMock {
...
-                                       passengers: passengers)
+        self.tripBooking = TripBooking(quoteId: tripBooking.quoteId)
         return self
     }
 
     func set(flightNumber: String) -> TripBookingMock {
         self.tripBooking = TripBooking(quoteId: tripBooking.quoteId,
-                                       passengers: tripBooking.passengers,
```
Comment (EdwardW2) [inline KarhooSDKTests/Mocks/Api/Request/TripBookingMock.swift:36]: Same

File: KarhooSDKTests/Mocks/Api/Request/TripBookingMock.swift [hunk 9]
```diff
@@ -15,32 +15,27 @@ final class TripBookingMock {
...
                                        flightNumber: flightNumber)
         return self
     }
     
     func set(meta: [String: Any]) -> TripBookingMock {
         self.tripBooking = TripBooking(quoteId: tripBooking.quoteId,
-                                       passengers: tripBooking.passengers,
```
Comment (EdwardW2) [inline KarhooSDKTests/Mocks/Api/Request/TripBookingMock.swift:43]: Same as Above

File: KarhooSDKTests/TestCases/Service/Trips/BookingInteractor/KarhooBookingInteractorSpec.swift [hunk 10]
```diff
@@ -120,7 +116,6 @@ class KarhooBookingInteractorSpec: XCTestCase {
      */
     func testRequestFails() {
         let tripBooking = TripBooking(quoteId: "some",
-                                      passengers: passengers,
```
Comment (EdwardW2) [inline KarhooSDKTests/TestCases/Service/Trips/BookingInteractor/KarhooBookingInteractorSpec.swift:123]: Yep, This file too, add all the Passengers back

File: KarhooSDKTests/TestCases/Service/Trips/KarhooTripServiceSpec.swift [hunk 11]
```diff
@@ -53,7 +53,7 @@ final class KarhooTripServiceSpec: XCTestCase {
      *   And:  Correct trip is returned in callback
      */
     func testBookingTripRequestSuccess() {
-        let tripBooking = TripBooking(quoteId: "some", passengers: mockPassengers)
```
Comment (EdwardW2) [inline KarhooSDKTests/TestCases/Service/Trips/KarhooTripServiceSpec.swift:56]: This file too

**Review by EdwardW2 — CHANGES_REQUESTED** [review]
Got quite a few changes to be made

File: KarhooSDK/Api/Request/TripBooking.swift (hunk 2 above)
Comment (nurito-burrito) [reply KarhooSDK/Api/Request/TripBooking.swift:14]: I will do that in the next ticket. I will add Passengers back then.

File: KarhooSDK/Api/Response/CancelledByPayer.swift (hunk 3 above)
Comment (nurito-burrito) [reply KarhooSDK/Api/Response/CancelledByPayer.swift:13]: Had to change the name as it was creating a conflict with a request file named Payer.

File: KarhooSDK/Api/Request/TripSearch.swift (hunk 1 above)
Comment (nurito-burrito) [reply KarhooSDK/Api/Request/TripSearch.swift:17]: Need more explanation on that

File: KarhooSDKTests/TestCases/Service/Trips/BookingInteractor/KarhooBookingInteractorSpec.swift (hunk 10 above)
Comment (nurito-burrito) [reply KarhooSDKTests/TestCases/Service/Trips/BookingInteractor/KarhooBookingInteractorSpec.swift:123]: done

File: KarhooSDKTests/Mocks/Api/Request/TripBookingMock.swift (hunk 9 above)
Comment (nurito-burrito) [reply KarhooSDKTests/Mocks/Api/Request/TripBookingMock.swift:43]: done

File: KarhooSDK/Api/Request/TripBooking.swift (hunk 2 above)
Comment (nurito-burrito) [reply KarhooSDK/Api/Request/TripBooking.swift:14]: added the missing ones.

File: KarhooSDKTests/TestCases/Service/Trips/KarhooTripServiceSpec.swift [hunk 12]
```diff
@@ -71,7 +71,7 @@ final class KarhooTripServiceSpec: XCTestCase {
      *  Then:  Error should be propagated through callback
      */
     func testBookingTripRequestFailure() {
-        let tripBooking = TripBooking(quoteId: "some", passengers: mockPassengers)
+        let tripBooking = TripBooking(quoteId: "some")
```
Comment (EdwardW2) [inline KarhooSDKTests/TestCases/Service/Trips/KarhooTripServiceSpec.swift:74]: Might want to return this one back to the way it was.

File: KarhooSDK/Api/Response/MeetingPoint.swift [hunk 13]
```diff
@@ -10,32 +10,38 @@ public struct MeetingPoint: Codable {
     public let position: Position
     public let instructions: String
     public let type: MeetingPointType
+    public let note: String
 
     public init(position: Position = Position(),
                 instructions: String = "",
-                type: MeetingPointType = .notSet) {
+                type: MeetingPointType = .notSet,
```
Comment (EdwardW2) [inline KarhooSDK/Api/Response/MeetingPoint.swift:17]: Doesn't need to be done now, but in MeetingPointType, make sure that Airport nor train station is listed. Can be done in Part 2.

File: KarhooSDKTests/TestCases/Service/Trips/KarhooTripServiceSpec.swift (hunk 12 above)
Comment (nurito-burrito) [reply KarhooSDKTests/TestCases/Service/Trips/KarhooTripServiceSpec.swift:74]: done

File: KarhooSDK/Api/Request/CancelReason.swift [hunk 14]
```diff
@@ -16,4 +16,7 @@ public enum CancelReason: String {
     case etaTooLong = "ETA_TOO_LONG"
     case driverDidntShowUp = "DRIVER_DIDNT_SHOW_UP"
     case otherUserReason = "OTHER_USER_REASON"
+    case foundBetterProce = "FOUND_BETTER_PRICE"
```
Comment (jsm20) [inline KarhooSDK/Api/Request/CancelReason.swift:19]: Typo as I think you mean foundBetterPrice?

**Review by jsm20 — DISMISSED** [review]

**Review by jsm20 — APPROVED** [review]

**Review by EdwardW2 — APPROVED** [review]

**Pull request merged**